sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,96 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
         shape ``(n_samples, K)``.
     """
     cdef Py_ssize_t i
-    cdef int32 node_id
     cdef Node *node
     for i in range(n_samples):
         node = root_node
         # While node not a leaf
-        while node.left_child != -1 and node.right_child != -1:
+        while node.left_child != TREE_LEAF:
             if X[i * n_features + node.feature] <= node.threshold:
                 node = root_node + node.left_child
             else:
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
 
-@cython.nonecheck(False)
+def _predict_regression_tree_stages_sparse(np.ndarray[object, ndim=2] estimators,
+                                           object X, double scale,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(<np.ndarray> X.data).data
```

Comment:
Shouldn't it be `<np.ndarray[ndim=1, dtype=DTYPE_t]>`?

Could you refer [`_apply_sparse_csr`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_tree.pyx#L814) and refactor it the same way. I feel for an additional 3 lines, it will be clearer... :)